### DISCUSSION TITLE
[ZK Stack] - Limit number of "eth_blockNumber" calls.

### DISCUSSION THREAD

**Discussion by sachanayush47** [zkSync-Community-Hub/zksync-developers#731] [ZK Stack] [answered]
### Module Affected

- [ ] System contracts (era-system-contracts)
- [ ] Smart contracts (era-contracts)
- [ ] zkEVM (era-zk_evm)
- [X] Server (zksync-era)
- [ ] Prover (era-boojum)
- [ ] Fee withdrawer (era-fee-withdrawer)
- [ ] Block Explorer (clock-explorer)
- [ ] Portal - Wallet + Bridge (dapp-portal)
- [ ] Other (please specify below)

### Rust Version

1.80.1

### Operating System

- [X] Ubuntu
- [ ] MacOS
- [ ] Windows
- [ ] Other (Please specify below)

### Issue Description

We have purchased Infura developer plan that offers 15 million credits per day. We have deployed a zksync hyperchain using Infura RPC. The problem is daily credit limit gets exhausted in less than 10 hours. Upon investigation we found out that more than 50% of credits is consumed by just `eth_blockNumber` method. Currently, we cannot move to higher plan as we still testing and figuring out things. We are trying to limit number of calls so that we can save on daily credit limit.
Only 5 - 10 transactions are made per day.

![Screenshot from 2024-10-14 16-05-52](https://github.com/user-attachments/assets/a8524b99-eb52-457d-bb4f-eb3cb01a0dce)

### Expected Behavior

 Some credits should be left at the end of the day

### Current Behavior

All the credits gets exhausted in less than 10 hours.

### Prior Research

- [X] Yes, I searched for similar issues and did not find any matching ones.

**Comment by alhenaai** [comment]
To address the issue of the `eth_blockNumber` method consuming a significant portion of your daily Infura credits, you can implement a caching mechanism on your server. This approach will reduce the frequency of calls made to the Infura RPC by storing the most recent block number and reusing it for a certain period before refreshing.

Here are some steps you can consider:

1. **Implement Caching**: Use a simple caching strategy where the `eth_blockNumber` result is stored and refreshed every few minutes instead of on every call. This can drastically reduce the number of requests made to Infura.

2. **Adjust the Refresh Rate**: Depending on the criticality of having the very latest block number, you might adjust the cache refresh rate. If your transactions are not time-sensitive, a longer cache duration could be feasible.

3. **Monitor Usage**: Keep an eye on the cache's effectiveness by monitoring how much it reduces the calls to `eth_blockNumber` and adjust the caching time accordingly.

4. **Optimize Further**: If possible, review other parts of your application to ensure that no unnecessary calls are made to the Infura RPC.

Implementing these changes should help in managing your Infura credits more efficiently and reduce the daily consumption caused by frequent `eth_blockNumber` calls.

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Comment by iscxyz** [comment, marked as answer]
Hello, 

you can monitor what component makes the most call via `server_ethereum_gateway_call` counters and `eth_client_direct` latency histograms. The relevant metrics are declared [here](https://github.com/matter-labs/zksync-era/blob/114834f357421c62d596a1954fac8ce615cfde49/core/lib/eth_client/src/clients/http/mod.rs#L40). Then you can reduce number of call by dialling down poll interval for respective component.

**Reply by sachanayush47** [reply]
> Hello,
> 
> you can monitor what component makes the most call via `server_ethereum_gateway_call` counters and `eth_client_direct` latency histograms. The relevant metrics are declared [here](https://github.com/matter-labs/zksync-era/blob/114834f357421c62d596a1954fac8ce615cfde49/core/lib/eth_client/src/clients/http/mod.rs#L40). Then you can reduce number of call by dialling down poll interval for respective component.

From `server_ethereum_gateway_call`, I found these components are making a lot of calls.
 What is  `watch` component? Im unable to find any config to increase the polling interval.


![image](https://github.com/user-attachments/assets/64572233-6883-4791-a0ee-0b724cd160b7)

**Reply by sachanayush47** [reply]
To change the intervals, increase the polling period in the `ecosystem/chains/chain_name/configs/general.yaml` file. This file contains the configuration for `ETH`, `gas adjuster`, and `watcher` settings.

These are the properties from `general.yaml` file which severely affect the number of requests made to the RPC.

`poll_period`
`eth_node_poll_interval`
`tx_poll_period`
`aggregate_tx_poll_period`
